### PR TITLE
Exact recon unet with batch normalization

### PR DESCRIPTION
**Pull request by kevinmicha**
Here the idea is to:

- Add batch normalization after every non-linearity
- Modify the ModelCheckpoint so that the model is saved every 500 epochs

### CODE REVIEW

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
Make sure to also include the `bn` flag in train/evaluation, and also in the unit test

Don't forget about the model checkpoint (I didn't see it here)

File: learning_wavelets/models/exact_recon_unet.py [hunk 1]
```diff
@@ -139,7 +153,8 @@ def call(self, inputs):
         outputs = self.final_conv(outputs)
         noise_std = tf.reshape(noise_std, shape=[tf.shape(noise_std)[0], 1, 1, 1])
         outputs = tf.image.resize_with_crop_or_pad(outputs, h, w)
-        outputs = noisy_image - noise_std * outputs
+        if self.exact_recon:
+            outputs = noisy_image - noise_std * outputs
```
Comment (zaccharieramzi) [inline learning_wavelets/models/exact_recon_unet.py:157]: Let's still make it residual
```suggestion
        if self.exact_recon:
            outputs = noisy_image - noise_std * outputs
        else:
        	outputs = noisy_image - outputs
```

File: learning_wavelets/tests/models/exact_recon_unet_test.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ def test_init():
     K.clear_session()
 
 def test_fit():
-    model=ExactReconUnet(n_output_channels=1, kernel_size=3, layers_n_channels=[4, 8])
+    model=ExactReconUnet(n_output_channels=1, kernel_size=3, layers_n_channels=[4, 8], bn=True, exact_recon=True)
```
Comment (zaccharieramzi) [inline learning_wavelets/tests/models/exact_recon_unet_test.py:14]: use `pytest` features to test all cases (see [here](https://docs.pytest.org/en/reorganize-docs/parametrize.html))

File: learning_wavelets/training_scripts/exact_recon_unet_training.py [hunk 3]
```diff
@@ -51,7 +53,7 @@ def train_unet(
 
     # callbacks preparation
 
-    chkpt_cback = ModelCheckpoint(chkpt_path, period=n_epochs, save_weights_only=True)
+    chkpt_cback = ModelCheckpoint(chkpt_path, period=500, save_weights_only=True)
```
Comment (zaccharieramzi) [inline learning_wavelets/training_scripts/exact_recon_unet_training.py:56]:
```suggestion
    chkpt_cback = ModelCheckpoint(chkpt_path, period=min(500, n_epochs), save_weights_only=True)
```

This way you won't do training that don't save their weights

File: learning_wavelets/training_scripts/exact_recon_unet_training.py [hunk 4]
```diff
@@ -80,7 +84,7 @@ def train_unet(
         steps_per_epoch=200,
         epochs=n_epochs,
         validation_data=im_ds_val,
-        validation_steps=1,
+        validation_steps=100,
```
Comment (zaccharieramzi) [inline learning_wavelets/training_scripts/exact_recon_unet_training.py:87]: I had forgotten that the validation dataset had a batch size already.
Let's make this smaller then
```suggestion
        validation_steps=15,
```

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
some remarks

File: learning_wavelets/models/exact_recon_unet.py [hunk 5]
```diff
@@ -139,7 +153,10 @@ def call(self, inputs):
...
+        else:
+        	outputs = noisy_image - outputs
```
Comment (zaccharieramzi) [inline learning_wavelets/models/exact_recon_unet.py:159]: watch out here: let's make sure we only use spaces and not tabulations for the whitespaces

File: learning_wavelets/models/exact_recon_unet.py [hunk 6]
```diff
@@ -171,13 +188,8 @@ def pad_power_of_two(x, n_layers):
...
     
-    new_w = w
-    new_h = h
-    if(w % power_of_two != 0):
-        new_w = power_of_two - w % power_of_two + new_w
-
-    if(h % power_of_two != 0):
-        new_h = power_of_two - h % power_of_two + new_h
+    new_w = power_of_two - w % power_of_two + w
+    new_h = power_of_two - h % power_of_two + h
```
Comment (zaccharieramzi) [inline learning_wavelets/models/exact_recon_unet.py:192]: Why did you make this change? this doesn't seem right if `w == power_of_two` for example

**Review by zaccharieramzi — APPROVED** [review]
Make you sure you re-write the training script for the old unet without click
Also, find a better name for the training script file

File: learning_wavelets/models/exact_recon_unet.py [hunk 7]
```diff
@@ -170,14 +187,14 @@ def pad_power_of_two(x, n_layers):
...
 
-    if(h % power_of_two != 0):
-        new_h = power_of_two - h % power_of_two + new_h
+    def not_pad_w(): return w
+    def pad_w(): return power_of_two - w % power_of_two + w
+    def not_pad_h(): return h
+    def pad_h(): return power_of_two - h % power_of_two + h
     
+    new_w = tf.cond(tf.equal(w % power_of_two, tf.zeros(1, tf.int32)), not_pad_w, pad_w)
+    new_h = tf.cond(tf.equal(h % power_of_two, tf.zeros(1, tf.int32)), not_pad_h, pad_h)
```
Comment (zaccharieramzi) [inline learning_wavelets/models/exact_recon_unet.py:197]: did you check that it worked?

**Review by zaccharieramzi — APPROVED** [review]
LGTM

File: learning_wavelets/tests/models/exact_recon_unet_test.py (hunk 2 above)
Comment (zaccharieramzi) [reply learning_wavelets/tests/models/exact_recon_unet_test.py:14]: you didn't solve this actually !

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
sorry about that, I had forgotten to verify the test. You didn't follow my advice on using `pytest` correctly

File: learning_wavelets/tests/models/exact_recon_unet_test.py (hunk 2 above)
Comment (kevinmicha) [reply learning_wavelets/tests/models/exact_recon_unet_test.py:16]: Now it's been uploaded

File: learning_wavelets/tests/models/exact_recon_unet_test.py [hunk 8]
```diff
@@ -10,12 +10,14 @@ def test_init():
     ExactReconUnet()
     K.clear_session()
 
-def test_fit():
-    model=ExactReconUnet(n_output_channels=1, kernel_size=3, layers_n_channels=[4, 8])
+@pytest.mark.parametrize("bn", [True, False])
+@pytest.mark.parametrize("exact_recon", [True, False])
+def test_fit(bn, exact_recon):
+    model=ExactReconUnet(n_output_channels=1, kernel_size=3, layers_n_channels=[4, 8], bn=True, exact_recon=True)
```
Comment (zaccharieramzi) [inline learning_wavelets/tests/models/exact_recon_unet_test.py:16]:
```suggestion
    model=ExactReconUnet(n_output_channels=1, kernel_size=3, layers_n_channels=[4, 8], bn=bn, exact_recon=exact_recon)
```

**Comment by zaccharieramzi** [comment]
@kevinmicha this looks good, but the CI tests are not running, can you look into why?

**Comment by kevinmicha** [comment]
> @kevinmicha this looks good, but the CI tests are not running, can you look into why?

done. Working now

File: learning_wavelets/models/exact_recon_unet.py [hunk 9]
```diff
@@ -170,14 +187,9 @@ def pad_power_of_two(x, n_layers):
...
-    new_w = w
-    new_h = h
-    if(w % power_of_two != 0):
-        new_w = power_of_two - w % power_of_two + new_w
 
-    if(h % power_of_two != 0):
-        new_h = power_of_two - h % power_of_two + new_h
-    
+    new_w = tf.cast(power_of_two*tf.math.ceil(w/power_of_two), tf.int32)
+    new_h = tf.cast(power_of_two*tf.math.ceil(h/power_of_two), tf.int32)
```
Comment (zaccharieramzi) [inline learning_wavelets/models/exact_recon_unet.py:192]: Excellent, can you change the documentation with this new approach (now it's not really related anymore)

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
Doc change

**Review by zaccharieramzi — APPROVED** [review]
LGTM
